feat(langgraph): support Pydantic state with aliased fields (input)

fede-kamel opened the pull request:
## Summary

Fixes #2555 — Pydantic `Field(alias=...)` and `alias_generator` are now respected on **both sides** of a LangGraph state schema: input and output.

### Before

```python
class State(BaseModel):
    foo: str = Field(alias='bar')

graph = StateGraph(State).compile()
graph.invoke({'bar': 'hello'})   # ValidationError (alias rejected)
graph.invoke({'foo': 'hello'})   # ValidationError (field name rejected on aliased field)
```

```python
class State(BaseModel):
    model_config = ConfigDict(
        alias_generator=alias_generators.to_camel,
        serialize_by_alias=True,
    )
    user_id: str
    display_name: str

graph.invoke({"userId": "u-1", "displayName": "Alice"})
# -> {"user_id": ..., "display_name": ...}    # output uses field names even though
                                              # schema says serialize_by_alias=True
```

### After

Both inputs and outputs respect the schema's alias intent.

---

## Input side

Channels are keyed by field name, but Pydantic rejects field-name kwargs for aliased fields. The fix translates between the two boundaries:

1. **`_get_pydantic_alias_map(schema)`** extracts `{alias: field_name}` from a Pydantic schema.
2. **`map_input` / `_get_updates`** translate user-supplied alias keys → channel (field) names before they reach a channel.
3. **`_coerce_state`** translates channel field names → aliases when constructing the Pydantic model (so Pydantic validation accepts them).

`alias_generator` is supported by construction: Pydantic populates `field_info.alias` at model-build time, so the same code path picks up generated aliases.

## Output side (new in `36d7ddec`)

Strictly opt-in via `model_config = ConfigDict(serialize_by_alias=True)` — mirroring Pydantic's own opt-in for alias-keyed `.model_dump()`. Graphs with aliased fields but without that flag are **unchanged** (no breakage).

When enabled, alias-keyed output flows through:

| Surface | Site |
|---|---|
| `invoke()` / `astream()` final return | `PregelLoop.output` (via `_rekey_with_aliases`) → `on_chain_end` |
| `stream(stream_mode='values')` | `map_output_values(output_alias_map=…)` |
| `stream(stream_mode='updates')` | `map_output_updates(output_alias_map=…)` |
| `get_state().values`, snapshot serialization | `Pregel._apply_output_aliases(...)` at both `StateSnapshot` construction sites |

The mechanism is a single tiny helper `_rekey_with_aliases(dict, map)` reused across the output surface, plus a `Pregel.output_alias_map` plumbed through `PregelLoop` from `StateGraph.compile`.

---

## Test plan

`libs/langgraph/tests/test_pydantic.py` (+4 new for the output side; +7 from the input commit):

**Input:**
- `test_pydantic_aliased_fields` — basic `Field(alias=...)`
- `test_pydantic_multiple_aliased_fields` — several explicit aliases
- `test_pydantic_alias_generator_to_camel` — canonical `alias_generator=to_camel`
- `test_pydantic_mixed_alias_and_fieldname` — mix of aliased and non-aliased
- `test_pydantic_alias_with_reducer` — aliased fields with reducer annotations
- `test_pydantic_aliased_fields_multinode_with_checkpointer` — multi-node + checkpointing + interrupt/resume
- `test_pydantic_aliased_fields_async` — async path

**Output (new):**
- `test_output_keys_use_alias_when_serialize_by_alias_set` — `invoke` + `stream(values)` emit alias keys
- `test_output_keys_unchanged_without_serialize_by_alias` — opt-in behaviour, no regression for non-opted-in graphs
- `test_stream_updates_use_alias_when_serialize_by_alias_set` — `stream(updates)` rekeys per-node update dicts
- `test_get_state_snapshot_values_use_alias_when_serialize_by_alias_set` — `StateSnapshot.values` uses aliases

```
tests/test_pydantic.py + tests/test_state.py: 33 passed (29 input + 4 output)
tests/test_pregel.py + tests/test_pregel_async.py (excluding redis/postgres variants): 890 passed, 0 regressions
make format / make lint / mypy:                                                       clean
```

The one failure in the broader suite (`test_in_one_fan_out_state_graph_waiting_edge_multiple[redis-True]`) reproduces on `origin/main` with identical content, so it's a pre-existing redis-cache flake, not introduced here.

spandan-sharma reviewed (commented):
We really need this too - `langgraph` currently completely ignores `alias` (even when `serialize_by_alias=True`) when creating its channels and uses whatever python-casing the pydantic-model field members had. The wire format (and channel names) should be alias names when provided.

fede-kamel reviewed (commented):
Thanks for the feedback! I considered naming channels by aliases, but chose field names because:

1. It matches Pydantic's design - Pydantic uses field names internally, aliases externally
2. Node functions use field names - when you access state.foo, that's the field name
3. Aliases are already supported at boundaries - input accepts aliases, output serializes with aliases
4. Backward compatible - existing code and checkpoints continue working

Channels are internal state management. The "wire format" (graph input/output) already uses aliases with this implementation.

Does this address your use case?

fede-kamel committed:
feat(langgraph): support Pydantic state with aliased fields

Fix #2555 — Pydantic Field(alias=...) and alias_generator are now
respected in LangGraph state schemas. Input dicts keyed by alias are
translated to channel field names, and _coerce_state rebuilds models
using aliases so Pydantic validation passes.

Covers simple aliases, alias_generator, mixed alias/non-alias fields,
reducers with aliases, checkpointer round-trips, and async invoke.

fede-kamel commented:
@spandan-sharma — `alias_generator` is supported. When Pydantic processes `ConfigDict(alias_generator=to_camel)`, it populates `field_info.alias` on each field. The implementation extracts these via `_get_pydantic_alias_map()`, so both explicit aliases and generated ones work:

```python
class State(BaseModel):
    model_config = ConfigDict(alias_generator=alias_generators.to_camel)
    first_name: str
    last_name: str

graph.invoke({"firstName": "John", "lastName": "Doe"})  # works
```

Note: the `channel_naming='alias'` output feature was removed to keep the PR focused on the core fix (input with aliases). Output wire-format consistency can be addressed separately if there's interest.

spandan-sharma commented:
> the channel_naming='alias' output feature was removed

@fede-kamel sorry I'm a bit confused - I thought we had discussed and agreed to include it and it was a part of this PR change-set when I had last checked. I also see a few of your previous PR comments are now missing/deleted. Did something happen in the meanwhile - why did we regress back to only partially fixing this when the code to fully fix it was already in place (AFAIK) ?

fede-kamel commented:
Happy to accommodate things back but I have not seen progress on this PR for months; neither traction from the maintainers  @spandan-sharma

fede-kamel committed:
test(langgraph): add alias_generator=to_camel coverage for #2555

The canonical Pydantic pattern for camelCase JSON APIs uses
alias_generator=to_camel with populate_by_name=True. The initial fix
covered Field(alias=...) explicitly but had no test demonstrating the
alias_generator path end-to-end, even though the implementation extracts
aliases via field_info.alias (which alias_generator populates).

fede-kamel commented:
@spandan-sharma — you're right, I reproduced the mixed-casing you described. With `alias_generator=to_camel` and `serialize_by_alias=True`:

```python
# stream_mode="values" / get_state() / invoke() returns:
{"channel_zero": "hi", "channel_one": {"fooZero": "a", "fooOne": 1}}
# outer: snake_case (channel names = field names)
# inner: camelCase (Pydantic serialize_by_alias)
```

Being honest about scope: this PR fixes the **input** side only. It unblocks the canonical #2555 use case (graph.invoke with camelCase payloads), but leaves the output wire format asymmetric as you noted. I've updated the PR title and description to reflect input-only scope.

On the output side: I think a `channel_naming="alias"` opt-in is the right design, but it touches checkpoint serialization and every streaming mode. That's enough surface area that I'd rather do it as a **dedicated follow-up PR** with its own review, than bundle it here and risk the scope killing this one too. I'm committing to opening that follow-up as soon as this lands — happy to draft it against the same issue.

Also added a test for `alias_generator=to_camel` (canonical Pydantic JSON-API pattern, e.g. `test_pydantic_alias_generator_to_camel`) since that's what most real-world users will actually hit.

@hinthornw @mdrxy — would appreciate your direction:
1. Merge as input-only fix, track output aliasing in a follow-up PR (my preference)
2. Bundle both into this PR

Either is fine by me. This PR has been open a few weeks without maintainer traction, so any signal helps. Thanks!

fede-kamel commented:
@sydney-runkle — fresh-eyes ping since you maintain both Pydantic and have merged langgraph PRs, and this is squarely in that intersection.

**What this PR does:** at the input boundary of a LangGraph state schema, respect Pydantic `Field(alias=...)` and `alias_generator`. Currently `graph.invoke({'bar': 'hello'})` raises `ValidationError` even when `State.foo = Field(alias='bar')`. Fix is +373/-12 across 6 files in `libs/langgraph/`, scoped to the input transform. Resolves #2555 (input side).

**State (refreshed today, `3967491`):**
- Rebased onto current main (was 93 commits behind, DIRTY). One real conflict in `pregel/main.py` — upstream added `StreamMessagesHandlerV2` to an import block I was also touching. Kept upstream's version (the new import is needed; my PR's content was incidental to that line).
- `make format` / `make lint` / `mypy` all clean
- `tests/test_pydantic.py` — 14/14 functional pass (3 `[postgres]`-variant tests skipped — need a local Postgres on port 5442 which the upstream CI provides via `make start-services`)

**The design question I'd like your call on:** @spandan-sharma asked (and previously got an in-flight implementation from @mdrxy that I later trimmed) for a complementary `channel_naming='alias'` flag that would also serialize *outputs* (`stream` / `get_state` / `invoke` return values) using aliases when `serialize_by_alias=True`. I deliberately scoped this PR to input only because:
1. The input fix is mechanically simpler and resolves the immediate crash for everyone using `alias_generator=to_camel`.
2. Output-side alias propagation touches channel naming and breaks the symmetry between checkpoint-stored keys vs wire-format keys — non-trivial design choice that I'd rather get reviewed separately.

If you'd prefer the unified scope, happy to fold the output-side work back in (the design is solid — I just didn't want to bundle two changes). If input-only is fine, this is ready as-is. **Either answer unblocks it.**

Top mergers (`jkennedyvz`, `longquanzheng`) haven't engaged; cc'ing you specifically because the alias semantics involve Pydantic decisions you'd weigh in on naturally.

fede-kamel committed:
feat(langgraph): also serialize state by alias on output (#2555)

The earlier commit on this branch added alias-aware *input* handling so
that `graph.invoke({'bar': 'hello'})` works when the state declares
`foo: str = Field(alias='bar')`. This commit completes the symmetry on
the output side: when the output schema sets
`model_config = ConfigDict(serialize_by_alias=True)`, all user-facing
output is keyed by alias rather than channel/field name.

Trigger:
  Strictly opt-in. The new behaviour kicks in only when the schema
  itself sets `serialize_by_alias=True` — matching Pydantic's own
  `.model_dump()` opt-in. Existing graphs with aliased fields but
  without that flag continue to emit field-name-keyed output (no
  behaviour change).

Mechanism:
  - `graph/state.py::_get_pydantic_output_alias_map(schema)` returns
    {field_name: alias} only when serialize_by_alias=True, else {}.
  - `StateGraph.compile()` computes this from output_schema and
    passes it through as Pregel.output_alias_map.
  - `Pregel._apply_output_aliases(values)` rekeys a dict at the
    StateSnapshot construction sites in both prepare paths.
  - `pregel/_io.py::_rekey_with_aliases(value, output_alias_map)` is
    the shared dict-rekey helper used by `map_output_values`,
    `map_output_updates`, and the loop's final `self.output =`
    sites (which feed `on_chain_end`).
  - `map_output_values` and `map_output_updates` take an
    `output_alias_map=None` keyword argument; the four `_emit` call
    sites in `_loop.py` pass `self.output_alias_map` through.

Tests (tests/test_pydantic.py, +4):
  - test_output_keys_use_alias_when_serialize_by_alias_set
  - test_output_keys_unchanged_without_serialize_by_alias
  - test_stream_updates_use_alias_when_serialize_by_alias_set
  - test_get_state_snapshot_values_use_alias_when_serialize_by_alias_set

Format / lint / mypy clean. 33/33 alias tests pass (29 existing + 4
new). Broader pregel suite: 890 passed, 0 regressions excluding a
pre-existing `[redis-True]` flake also present on origin/main.

Addresses @spandan-sharma's output-side feedback on this PR.

fede-kamel commented:
Pushed `36d7ddec` adding the output-side fix that @spandan-sharma asked about.

**TL;DR:** the PR now covers both sides of the alias question, not just input.

When the output schema sets `model_config = ConfigDict(serialize_by_alias=True)` — same opt-in Pydantic uses for `.model_dump()` — `invoke` returns, `stream(values)` / `stream(updates)` chunks, `StateSnapshot.values` from `get_state`, and the `on_chain_end` callback payload all emit keys using the declared aliases instead of channel/field names. Schemas without `serialize_by_alias=True` are completely unchanged — no behaviour shift, no breakage.

**Mechanism:** one helper `_rekey_with_aliases(dict, map)` in `pregel/_io.py`, threaded as `Pregel.output_alias_map` from `StateGraph.compile` through `PregelLoop` to the 4 `_emit` call sites and the 2 `StateSnapshot` construction sites. Compact diff (5 files, +299/-13).

**Tests** (`tests/test_pydantic.py`):
- ✅ `test_output_keys_use_alias_when_serialize_by_alias_set`
- ✅ `test_output_keys_unchanged_without_serialize_by_alias` ← explicit opt-in guard against regressions
- ✅ `test_stream_updates_use_alias_when_serialize_by_alias_set`
- ✅ `test_get_state_snapshot_values_use_alias_when_serialize_by_alias_set`

33/33 alias tests pass. Broader pregel suite: 890 passed (1 pre-existing `[redis-True]` flake also fails on `origin/main`). `make format` / `make lint` / `mypy` all clean.

@spandan-sharma — the `channel_naming='alias'` semantics you were after are in: enabled by `serialize_by_alias=True` on the schema rather than a separate flag, which keeps the API surface aligned with what Pydantic itself does. Would appreciate your eyes on the test for the `stream(values)` mixed-casing case you raised earlier.

@sydney-runkle — the choice to gate on `serialize_by_alias` (rather than a new `channel_naming` parameter) is to match Pydantic semantics by construction. Happy to switch to an explicit flag if you'd prefer.

fede-kamel commented:
@spandan-sharma — closing this aliased-Pydantic-state proposal. Your feedback correctly identified that input, output, streaming, and snapshot semantics need to be designed as one coherent wire-format decision. The branch is now stale and conflicts with current main, so it is no longer responsible to keep it presented as merge-ready. The underlying issue retains the problem statement and design discussion. Thank you for the careful review.
